File: Data Structures/4. Swift/j. HashTables.swift

struct HashTable<Key: Hashable, Value> {
    private typealias Element = (key: Key, value: Value)
    private typealias Bucket = [Element]

    private var buckets: [Bucket]

    private var totalBuckets: Int {
        return buckets.count
    }

    // MARK: - Initialization

    init(capacity: Int) {
        assert(capacity > 0, "Capacity should be greater than 0")
        self.buckets = Array(repeating: [], count: capacity)
    }

    // MARK: - Access

    /// Get the value associated with the given key.
    func getValue(forKey key: Key) -> Value? {
        let index = bucketIndex(for: key)
        for element in buckets[index] where element.key == key {
            return element.value
        }
        return nil
    }

    /// Set the value for the given key.
    mutating func setValue(_ value: Value, forKey key: Key) {
        let index = bucketIndex(for: key)

        // Check if the key already exists, and update the value
        for (i, element) in buckets[index].enumerated() {
            if element.key == key {
                buckets[index][i].value = value
                return
            }
        }

        // Key doesn't exist, add a new entry
        buckets[index].append((key: key, value: value))
    }

    // MARK: - Removal

    /// Remove the value associated with the given key.
    mutating func removeValue(forKey key: Key) {
        let index = bucketIndex(for: key)
        buckets[index] = buckets[index].filter { $0.key != key }
    }

    /// Remove all elements from the hash table.
    mutating func removeAll() {
        buckets = Array(repeating: [], count: totalBuckets)
    }

    // MARK: - Helpers

    /// Get the index of the bucket for the given key.
    private func bucketIndex(for key: Key) -> Int {
        return abs(key.hashValue) % totalBuckets
    }

    // MARK: - Count

    /// Get the total count of elements in the hash table.
    func count() -> Int {
        return buckets.reduce(0) { $0 + $1.count }
    }

    // MARK: - Check Existence

    /// Check if the hash table contains a value for the given key.
    func contains(key: Key) -> Bool {
        let index = bucketIndex(for: key)
        return buckets[index].contains { $0.key == key }
    }

    // MARK: - Keys and Values

    /// Get an array of all keys in the hash table.
    func allKeys() -> [Key] {
        let keys = buckets.flatMap { $0.map { $0.key } }
        return keys
    }

    /// Get an array of all values in the hash table.
    func allValues() -> [Value] {
        let values = buckets.flatMap { $0.map { $0.value } }
        return values
    }

    // MARK: - Merging

    /// Merge the contents of another hash table into this hash table.
    mutating func merge(with otherTable: HashTable<Key, Value>) {
        for bucket in otherTable.buckets {
            for element in bucket {
                setValue(element.value, forKey: element.key)
            }
        }
    }

    // MARK: - Resizing

    /// Resize the hash table to the new capacity.
    mutating func resize(to newCapacity: Int) {
        // Create a new hash table with the desired capacity
        var newHashTable = HashTable<Key, Value>(capacity: newCapacity)

        // Merge the elements from the current hash table
        newHashTable.merge(with: self)

        // Assign the new hash table
        self = newHashTable
    }

    // ... Add more hash table operations as needed ...
}

File: data_structures/6_go/j_hashTables.go
package main

import (
	"fmt"
	"hash/fnv"
)

type HashTable struct {
	capacity int
	buckets  [][]pair
}

type pair struct {
	key   interface{}
	value interface{}
}

func NewHashTable(capacity int) *HashTable {
	if capacity <= 0 {
		panic("Capacity should be greater than 0")
	}

	buckets := make([][]pair, capacity)
	for i := range buckets {
		buckets[i] = []pair{}
	}

	return &HashTable{
		capacity: capacity,
		buckets:  buckets,
	}
}

func (ht *HashTable) bucketIndex(key interface{}) int {
	h := fnv.New32a()
	fmt.Fprintf(h, "%v", key)
	hash := h.Sum32()
	return int(hash) % ht.capacity
}

func (ht *HashTable) GetValue(key interface{}) interface{} {
	index := ht.bucketIndex(key)
	for _, p := range ht.buckets[index] {
		if p.key == key {
			return p.value
		}
	}
	return nil
}

func (ht *HashTable) SetValue(key interface{}, value interface{}) {
	index := ht.bucketIndex(key)

	// Check if the key already exists, and update the value
	for i, p := range ht.buckets[index] {
		if p.key == key {
			ht.buckets[index][i] = pair{key, value}
			return
		}
	}

	// Key doesn't exist, add a new entry
	ht.buckets[index] = append(ht.buckets[index], pair{key, value})
}

func (ht *HashTable) RemoveValue(key interface{}) {
	index := ht.bucketIndex(key)

	// Remove the key if it exists
	var newBucket []pair
	for _, p := range ht.buckets[index] {
		if p.key != key {
			newBucket = append(newBucket, p)
		}
	}
	ht.buckets[index] = newBucket
}

func (ht *HashTable) RemoveAll() {
	for i := range ht.buckets {
		ht.buckets[i] = []pair{}
	}
}

func (ht *HashTable) Count() int {
	count := 0
	for _, bucket := range ht.buckets {
		count += len(bucket)
	}
	return count
}

func (ht *HashTable) Contains(key interface{}) bool {
	index := ht.bucketIndex(key)
	for _, p := range ht.buckets[index] {
		if p.key == key {
			return true
		}
	}
	return false
}

func (ht *HashTable) AllKeys() []interface{} {
	var keys []interface{}
	for _, bucket := range ht.buckets {
		for _, p := range bucket {
			keys = append(keys, p.key)
		}
	}
	return keys
}

func (ht *HashTable) AllValues() []interface{} {
	var values []interface{}
	for _, bucket := range ht.buckets {
		for _, p := range bucket {
			values = append(values, p.value)
		}
	}
	return values
}

func (ht *HashTable) Merge(otherTable *HashTable) {
	for _, bucket := range otherTable.buckets {
		for _, p := range bucket {
			ht.SetValue(p.key, p.value)
		}
	}
}

func (ht *HashTable) Resize(newCapacity int) {
	if newCapacity <= 0 {
		panic("New capacity should be greater than 0")
	}

	newBuckets := make([][]pair, newCapacity)
	for i := range newBuckets {
		newBuckets[i] = []pair{}
	}

	for _, bucket := range ht.buckets {
		for _, p := range bucket {
			newIndex := ht.bucketIndex(p.key)
			newBuckets[newIndex] = append(newBuckets[newIndex], p)
		}
	}

	ht.buckets = newBuckets
}

func (ht *HashTable) PrintHashTable() {
	for i, bucket := range ht.buckets {
		fmt.Printf("Bucket %d: ", i)
		for _, p := range bucket {
			fmt.Printf("(%v, %v) ", p.key, p.value)
		}
		fmt.Println()
	}
}
